Proper divisor functions

adhoc-king opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #15927 

#### Brief description of what is fixed or changed
A proper divisor function was missing. Added function for proper divisor and its count

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- ntheory
    - Added proper_divisor and proper_divisor_count functions
<!-- END RELEASE NOTES -->

Commit:
Update factor_.py

Commit:
Update factor_.py

Commit:
Update factor_.py

Commit:
Update factor_.py

oscargus commented:
I do not know the theory behind it, so I cannot comment on that or the code, put please add a few tests.

smichr commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1419,20 @@ def divisor_count(n, modulus=1):
+        return divisors(n)[1:-1]
+    else:
+        x = divisors(n, True)
+        x.remove(1)
```
Does a generator have a remove method? I suspect you want to do something like

```
for i in divisors(n, generator=True):
    if i > 1 and i < n:
        yield i
```

adhoc-king replied:
How about this?
```python
x = list(divisors(n, True))
```

asmeurer replied:
That defeats the purpose of making it a generator. 

1 is always the first thing returned by divisors, so you just need to call `next(x)` to get rid of it. That doesn't get rid of the last item, though, so I think you will need a separate function as @smichr suggests (it needs to be separate so that `generator=True` works). 

Unless there is something in itertools or sympy.utilities that could do this for you. I thought maybe [islice](https://docs.python.org/3/library/itertools.html#itertools.islice) but it doesn't let you use a negative slice for the stop.

adhoc-king replied:
I think the code by @smichr is quite perfect.

Commit:
Update factor_.py

asmeurer commented:
This needs tests.

asmeurer commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1419,36 @@ def divisor_count(n, modulus=1):
     return Mul(*[v + 1 for k, v in factorint(n).items() if k > 1])
 
 
+def proper_divisors(n, generator=False):
```
Add a docstring for this function.

adhoc-king replied:
Done :)

asmeurer commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1419,36 @@ def divisor_count(n, modulus=1):
+    See Also
+    ========
+
+    divisors, proper_divisors, divisors_count
```
Should be

```suggestion
    divisors, proper_divisors, divisor_count
```

adhoc-king replied:
Thanks! Done

Commit:
Update factor_.py

Commit:
Update factor_.py

asmeurer commented:
Please write informative commit message for your changes (see https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages). I recommend learning to use git and making the changes locally on your computer, rather than in the GitHub web interface (that document goes over how to do that).

Commit:
Update factor_.py

Fixed proper_divisors function

asmeurer commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1419,53 @@ def divisor_count(n, modulus=1):
+    factorint, divisors, proper_divisor_count
+    
+    """
+    for i in divisors(n, generator):
```
This does not not give a generator if `generator=False`. You will need to have a separate internal function that is a generator that gets returned when `generator=True`.

asmeurer commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1419,53 @@ def divisor_count(n, modulus=1):
+    
+    """
+    for i in divisors(n, generator):
+        if i > 1 and i < n:
```
You can use 
```suggestion
        if 1 < i < n:
```

asmeurer committed:
Update sympy/ntheory/factor_.py

Minor change

Co-Authored-By: adhoc-king <46354827+adhoc-king@users.noreply.github.com>

Commit:
Update factor_.py

Fixed generator of proper_divisor function

Commit:
Update factor_.py

Fixed bug: proper divisors includes 1

Commit:
Update factor_.py

adhoc-king commented:
I can't believe that I forgot that proper divisors include 1, have fixed it.

Commit:
Update factor_.py

Added an internal function for generator

Commit:
Update factor_.py

Added some cases for optimization and support for negative numbers

Commit:
Update test_factor_.py

Added tests for proper_divisors() and proper_divisor_count( functions

Commit:
Update __init__.py

Imported proper_divisors and proper_divisor_count

Commit:
Update factor_.py

Removed trailing whitespaces

Commit:
Update factor_.py

Removed trailing whitespaces

Commit:
Update test_factor_.py

Removed trailing whitespaces

Commit:
Update test_factor_.py

Added proper_divisors and proper_divisor_count

smichr commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1419,82 @@ def divisor_count(n, modulus=1):
     return Mul(*[v + 1 for k, v in factorint(n).items() if k > 1])
 
 
+def proper_divisors(n, generator=False):
```
Let's give `_divisors` a 2nd argument, `proper=False` and call `_divisors` from `proper_divisors` rather than duplicating code:
```python
def _divisors(n, proper=False):
    ...
    if not proper or p != n: yield p
```

adhoc-king replied:
How about this?
```python
if proper:
    for p in rec_gen():
        if p != n:
            yield p
else:
    for p in rec_gen():
        yield p
```

oscargus replied:
And a generator argument as well? Seemed like a neat idea (but I haven't checked the code so I cannot tell how feasible it is).

oscargus replied:
I think I didn't notice the _...

adhoc-king replied:
@smichr Added ```proper``` argument. Thanks!

Commit:
Update factor_.py

Fixed bug when proper_divisor_count is called for n = 0

Commit:
Update factor_.py

Added proper argument in _divisors() helper function

asmeurer commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1424,63 @@ def divisor_count(n, modulus=1):
+    if n == 0:
+        return []
+
+    pd = _divisors(n, True)
```
I would pass this explicitly as a keyword argument. 

```suggestion
    pd = _divisors(n, proper=True)
```

Once we drop Python 2 support I think we should make "option" keyword arguments like these keyword-only arguments.

adhoc-king replied:
Thanks!

asmeurer committed:
Update sympy/ntheory/factor_.py

Explicit keyword argument

Co-Authored-By: adhoc-king <46354827+adhoc-king@users.noreply.github.com>

adhoc-king commented:
All checks have passed. If there are no other suggestions then this can be merged now

smichr commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1424,63 @@ def divisor_count(n, modulus=1):
+    factorint, divisors, proper_divisor_count
+
+    """
+    n = as_int(abs(n))
```
You are duplicating code here, too...so how about giving divisors a 3rd argument `proper=False` and then your `proper_divisors` is just a thin wrapper to `return divisors(n, generator, proper=True)`.

adhoc-king replied:
@smichr Done, thanks! Please check it again

Commit:
Update factor_.py

Remove duplicate code in proper_divisor() by replacing it with call to divisor() function

oscargus commented on sympy/ntheory/factor_.py:
```diff
@@ -1419,6 +1424,51 @@ def divisor_count(n, modulus=1):
+    factorint, divisors, proper_divisor_count
+
+    """
+    return divisors(n, generator, proper=True)
```
Minor thing, but for consistency I would go for either:
`return divisors(n, generator=generator, proper=True)`
or
`return divisors(n, generator, True)`
with a preference for the former.

oscargus commented:
Looks good! Just a minor comment.

Commit:
Update factor_.py

Minor bug fix for n=1

smichr committed:
Update factor_.py

smichr commented:
> Just a minor comment.

I did that and cancelled the previous test. Could you please squash this into a single commit and then force push here?

smichr commented:
When you force push here, the tests will be restarted.

Commit:
Update factor_.py

smichr commented:
please do the following locally:

```
git reset head~27 {or how many commits there have been}
{then commit with a title like "add proper_divisors and proper_divisors_count"}
git push -f adhoc-king
```

smichr committed:
divisor_count gets proper keyword

And proper_divisor_count gets modulus keyword.

smichr commented:
This is just waiting for a clean-up of commit history.

jhanwar commented:
As it is not merged yet. I opened a new pull request.

czgdp1807 commented:
@adhoc-king @smichr @asmeurer @oscargus Any news on this? Can we conclude this PR?

smichr commented:
> Can we conclude this PR?

https://github.com/sympy/sympy/pull/15928#issuecomment-461593090

czgdp1807 commented:
I am continuing this then.

czgdp1807 commented:
Closing this in favour of https://github.com/sympy/sympy/pull/18085
Thanks for the contributions.
